Document end-to-end formulations with GEMD example (#586)
CitrineInformatics/citrine-python: docs/source/formulations_example.rst

.. formulations_example:

Formulations Example
====================

The Citrine Platform is a powerful tool to analyze formulations data.
A formulation is, generally speaking, a mixture of mixtures.
Individual ingredients are mixed together to form more complex ingredients, which may themselves be mixed with other ingredients and/or mixtures, and so on.

> Generally, the top-level formulation has some properties that we with to optimize.

Reviewer: ```suggestion
Generally, the top-level formulation has some properties that we wish to optimize.
```

> The ingredients themselves may also have known properties, which are useful for understanding the properties of the top-level formulation.

Reviewer: ```suggestion
The ingredients themselves may also have known properties that are useful for understanding the properties of the top-level formulation.
```

In addition, each mixing process may be characterized by attributes such as the temperature of a reaction vessel or the speed of a mixing element.

The GEMD data model provides a way to record all of the information relevant to a formulations problem,
and Citrine's AI Engine turns this information into trained models that can predict the properties of novel formulations.
This section demonstrates how to express a complete formulations example in Citrine-python, from data ingestion to new candidate generation.

Example Raw Data
----------------

To demonstrate the formulations machinery, we consider a toy problem -- margaritas!
We scour our pantry and gather some ingredients, laid out in the table below.
Several ingredients have labels, which helps the Citrine Platform understand the relationships between ingredients.

> We have also measured the sugar fraction for some of the ingredients, and we know the price per kilogram of each ingredient.

Reviewer 2: MAYBE - Could we use something else besides $/kg? This is relevant to Materials as well, but we don't support this unit natively in GEMD (AFAIK) and may set poor expectations.

Reviewer: Concur.  Perhaps trying to hit a particular alcohol content?

Author: In the actual example, I use `1/kg` as the unit.

Constraining on a mean property is such a common use case that it should be demonstrated, and limiting price seems to be the most common manifestation. But if you think that the unit issue is too thorny, then I agree that "alcohol content" is another example that could demonstrate the concept.

Author: I added a note about the currency unit

     - 0.50
     - 0.50
   * - Simple Syrup B
     - Simple Syrup, Sweetener
     - 0.45
     - 0.55

We now create margaritas by mixing together a subset of ingredients and blending for some time.
The basic ingredients are labeled as described in the "Ingredients Data" table, and the simple syrups are given a "simple syrup" label.
In addition to the ingredient mass fractions, we record the blending time, in seconds.
The margaritas are judged on an ultra-rigorous "tastiness" scale, which ranges from 0.0 to 10.0
The table below shows two examples.

.. list-table:: Margarita Data
   :widths: 35 25 40 35 35 35 35 25 25 25
   :header-rows: 1

   * - Name
     - Tastiness
     - Blend Time (s)
     - Simple Syrup A Quantity
     - Simple Syrup B Quantity
     - Sugar Quantity
     - Lime Juice Quantity
     - Triple Sec Quantity
     - Tequila Quantity
     - Ice Quantity
   * - Margarita A
     - 6.3 +/- 0.1
     - 8.0
     - 0.20
     - 0.0
     - 0.0
     - 0.15
     - 0.0
     - 0.25
     - 0.40
   * - Margarita B
     - 5.4 +/- 0.1
     - 12.0
     - 0.0
     - 0.15
     - 0.0
     - 0.10
     - 0.10
     - 0.30
     - 0.35

Ingesting Data
--------------

> .. Warning:: Ingesting data to GEMD is best done with additional tooling to automate the process.

Reviewer: ```suggestion
.. Warning:: Ingesting data to GEMD is most easily done with additional tooling to automate the process.
```

Although we will not describe every step of the data ingestion process, we will highlight several data objects and their inter-connections.

> Perhaps most crucial are the `attribute templates` that correspond to the parameters of the blending process and the properties of the ingredients/formulations.

Reviewer: ```suggestion
Perhaps most crucial are the `attribute templates`, which correspond to the parameters of the blending process and the properties of the ingredients/formulations.
```

.. note:: In the code examples that follow, imports are assumed to be cumulative.
    Once an import has been demonstrated, we do not include it in subsequent code blocks.

.. code-block:: python

    from citrine.resources.parameter_template import ParameterTemplate
    from citrine.resources.property_template import PropertyTemplate
    from gemd.entity.bounds import RealBounds

    # Assume we have already created a project and dataset, named "project" and "dataset."
    # Though not demonstrated here, it is useful to give templates custom ids so that they can be easily retrieved.
    blend_time_template = dataset.parameter_templates.register(
        ParameterTemplate("blend time", bounds=RealBounds(0, 60.0, "s"))
    )
    price_template = dataset.property_templates.register(
        PropertyTemplate("price", bounds=RealBounds(0, 100, "1/kg"))
    )
    sucrose_fraction_template = dataset.property_templates.register(
        PropertyTemplate("sucrose fraction", bounds=RealBounds(0, 1, ""))
    )
    tastiness_template = dataset.property_templates.register(
        PropertyTemplate("tastiness", bounds=RealBounds(0, 10, ""))

>     )

Reviewer: Given that you registering these, would it be better to assign UIDs and use LinkByUID for reference, rather than object references?

Author: yes, it would

The attribute templates are attached to relevant object templates.
For example, the process template to represent blending should include ``blend_time_template`` as a ``parameter``.

Although it contains no attributes, we should particular pay attention to the templates that represent the mixing processes.
These will be used to convert GEMD data into formulations.
It is possible to use different process templates to distinguish between different types of mixing, but here we use the one template to represesnt both types of mixing that occur (mixing the simple syrup and mixing the margarita ingredients).

> The template includes a comprehensive list of all allowed names and labels.

Reviewer 2: SHOULD - Clarify that we mean _ingredient_ names here. Maybe:
> ...includes a comprehensive list of all allowed ingredient names and ingredient labels that can go into the process.

Often it's not possible to enumerate all the ingredient names/labels when making process templates. It may be good to call out that this is not a requirement for GEMD, but will make creating tables later on much easier.

.. code-block:: python

    from citrine.resources.process_template import ProcessTemplate

    mix_template = dataset.process_templates.register(
        ProcessTemplate(
            "mix"

>             allowed_names=["simple syrup", "sugar", "water", "lime juice", "triple sec", "tequila", "ice"],

Reviewer: `lime juice` and `triple sec` might be more specific than we want.  At the least, I'd swap to `orange liqueur`.

            allowed_labels=["sweetener", "acid", "alcohol", "simple syrup"]
        )
    )

Notice that "simple syrup" is an allowed ingredient name, but we do not distinguish between "simple syrup A" and "simple syrup B."

> This reflects the fact that we only ever use one simple syrup in a mixing process.

Reviewer 2: SHOULD - explain why this is. E.g. "... as GEMD does not allow two ingredients of the same name to go into a single Process Object."

The resulting table will have one column indicating which simple syrup was used and a second column indicating how much was used.

> Distinguishing in this way between the unique id of the material ("simple syrup A," "simple syrup B," etc.) and the generic name of the ingredient ("simple syrup") is especially useful when there are many materials to choose from.

Reviewer 2: MAYBE - I wonder if we should have a break-out box on naming ingredients. This is a common way that people build debt in their GEMD structure. I've been documenting some of the paths forward for DS/DE [here](https://docs.google.com/document/d/1oDlFkXcO570T7yJM1E_xUs32Le1VEC9ssM0o9w7qjA4/edit#heading=h.5a9yl832y301). 

The main points we might want to get across is that:
* The Material UID will correspond to the most specific quantity you can use
* The Ingredient Name will correspond to how ingredients are grouped into columns
* The Ingredient Label(s) will allow grouping of multiple Material UID's when building models and design spaces.

    mix_margarita_spec = dataset.process_specs.register(
        ProcessSpec(f"mix margarita B", template=mix_template)
    )
    dataset.ingredient_specs.register(
        IngredientSpec(
            "simple syrup",
            material=simple_syrup_B_spec,
            process=mix_margarita_spec,
            labels=["simple syrup"]
            mass_fraction=NominalReal(nominal=0.15, units="")
        )
    )
    dataset.ingredient_specs.register(
        IngredientSpec(
            "triple sec",
            material=triple_sec_spec,
            process=mix_margarita_spec,
            labels=["alcohol", "sweetener"],
            mass_fraction=NominalReal(nominal=0.1, units="")
        )
    )
    # register remaining ingredient specs....
    # Then register the resulting material spec.
    margarita_spec = dataset.material_specs.register(
        MaterialSpec(f"margarita B", process=mix_margarita_spec, template=margarita_template)
    )

This material spec is then fed as the sole ingredient into a "blend margarita B" process spec, which produces a "blended margarita B" material spec.
A measurement spec is attached to the material spec to measure "tastiness."

> Finally, run objects are created corresponding to each spec, to represent what actually happened.

Reviewer: Can we get an example?

Author: I'd rather not. Since ingesting GEMD is not the point of this example, I only want to include the minimum information that is necessary to make the rest of the tutorial make sense.

Whew!
That's a lot, which is why this is best done with additional tooling, whether in code or in the GUI.

A rendering of this example material history is shown below.

.. figure:: _static/GEMD_history_example.png
    :align: center

    Material History for Blended Margarita B

Repeating this process once for each margarita sample, we can build up a rich dataset for machine learning.

Building a Table
----------------

We now build a GEM Table to represent the margaritas' material histories in tabular format.
This table will be used as training data when building a machine learning graphical model.
For more detailed information on GEM Tables, see the section on :doc:`data extraction <data_extraction>`.

> In order to make a Gem Table, we start with a Table Configuration object.

Reviewer: ```suggestion
In order to make a GEM Table, we start with a Table Configuration object.
```

In this example we will build up the configuration in small steps.
As we will see, templates are crucial to configuring the table.

The code below defines the rows and defines one column that contains the identifier of each row.

.. code-block:: python

    from citrine.gemtables.rows import MaterialRunByTemplate
    from citrine.gemtables.variables import RootIdentifier
    from citrine.gemtables.columns import IdentityColumn
    from gemd.entity.link_by_uid import LinkByUID

    material_templates_to_include = [
        blended_margarita_template, simple_syrup_template, base_material_template
    ]
    scope = "margaritas-id"

    config = TableConfig(
        name="margarita formulations table",
        description="",
        datasets=[dataset.uid],
        rows=[
            MaterialRunByTemplate(
                templates=[LinkByUID.from_entity(t) for t in material_templates_to_include]
            )
        ]
        variables=[RootIdentifier(name="name", headers=["name"], scope=scope)],
        columns=[IdentityColumn(data_source="name")]
    )

Let's step through the pieces of this code.

> We gave the configuration a name and description, for human-readability.

Reviewer: ```suggestion
We gave the configuration a name and description for human-readability.
```

We defined the datasets that contain the material histories; for the purposes of this example we assume everything is in the dataset ``dataset``.

> We then defined the rows as being based on all materials that link to one of a set of templates.

Reviewer: ```suggestion
We then defined the rows as being based on all materials that link to one of a set of Material Templates.
```

For the purposes of this example, assume that we have defined separate material templates for the base materials, the simple syrups, the margaritas, and the blended margaritas.
Notice that we do _not_ include the material template for the unblended margaritas.
For the purposes of machine learning, we want to compress the mixing and blending into a single training row.

We define a single variable/column that contains a unique identifier for each row.
Every object on the Citrine Platform has a unique id with scope ``id``, but these are difficult for a human to read.

> In this example we assume that we have given the materials human-readable uniquely identifying names with scope "margaritas-id."

Reviewer: So why didn't we do that above?  There's no demonstration of assigning a UID.

Author: I added custom ids

        name=f"{ing_name} mass",
        headers=[ing_name, "Mass"],
        process_templates=[mix_template_link],
        ingredient_name=ing_name,
        quantity_dimension=IngredientQuantityDimension.MASS,
    )
    labels_var = IngredientLabelsSetInOutput(
        name=f"{ing_name} labels",
        headers=[ing_name, "Labels"],
        process_templates=[mix_template_link],
        ingredient_name=ing_name
    )
    config = config.add_columns(
        variable=identity_var, columns=[IdentityColumn(data_source=identity_var.name)]
    )
    config = config.add_columns(
        variable=quantity_var, columns=[MeanColumn(data_source=quantity_var.name)],
    )
    config = config.add_columns(
        variable=labels_var,
        columns=[
            ConcatColumn(
                data_source=labels_var.name,
                subcolumn=IdentityColumn(data_source=labels_var.name)
            )
        ]
    )

The figure below shows how these three columns may be rendered in a GEM Table.

.. figure:: _static/GEM_Table_formulation_ingredient.png
    :align: center

    GEM Table "simple syrup ingredient" columns

Lastly, we register the configuration and build the table.

> Note that this is a long-running process.

Reviewer 2: SHOULD - "... is often a long-running..." Some demos only take a few seconds.

Reviewer: ```suggestion
Note that this can be a long-running process depending on how many cells are in the table.
```

To build the table asynchronously, use :func:`~citrine.resources.gemtables.GemTableCollection.initiate_build`.

.. code-block:: python

    config = project.table_configs.register(config)
    table = project.tables.build_from_config(config)

Training a Predictor
--------------------

With the GEM Table in hand, we build and train a predictor to predict the tastiness of novel margarita recipes.

> The first step is to define a :class:`~citrine.informatics.data_sources.GemTableDataSource` based on the Gem Table, ``table``.

Reviewer: ```suggestion
The first step is to define a :class:`~citrine.informatics.data_sources.GemTableDataSource` based on the GEM Table, ``table``.
```

We choose to define a :class:`~citrine.informatics.descriptors.FormulationDescriptor` to hold the formulation;
if we do not specify it then a default descriptor will be generated, but given how crucial this descriptor is it is best to specify it directly.

.. code-block:: python

    from citrine.informatics.descriptors import FormulationDescriptor
    from citrine.informatics.data_sources import GemTableDataSource

    formulation = FormulationDescriptor("mixed and blended margarita")
    data_source = GemTableDataSource(table_id=table.uid, table_version=table.version, formulation_descriptor=formulation)

> The first component of the graphical model is a :class:`~citrine.informatics.predictors.simple_mixture_predictor.SimpleMixturePredictor`, which flattens the input formulation--it repeatedly replaces components with their ingredients until only the basic ingredients remain.

Reviewer 2: MAYBE - Replace "basic" with "atomic". If we can keep this language throughout the documentation that would be great. Atomic ingredient = one that can't be broken down any further. "Basic" could mean a lot of things.

Author: I wasn't sure what external-facing language we standardized on, but I like atomic

This flattening efficiently teaches the predictor about the relationship between ingredients.

> In this case, it learns exactly how "simple syrup A" and "simple syrup B" are similar to each other.

Reviewer 2: MAYBE - "each other by reducing them to their atomic ingredients: water and sugar." 

I'm not sure how ubiquitous the term "flattening" is - it might be helpful to users to explain what's going on a couple of different ways.

Although the homogeneous representation is not entirely appropriate for all formulations problems, it is usually an excellent approximation,

> especially when coupled with flexible machine learning models that can learn subtle relationships within the data.

Reviewer: ```suggestion
especially when coupled with flexible machine learning models that can emulate more complex relationships within the data.
```

.. code-block:: python

    from citrine.informatics.predictors import SimpleMixturePredictor

    flat_formulation = FormulationDescriptor("homogenized margarita")
    simple_mixture_predictor = SimpleMixturePredictor(
        name="Simple margarita mixture",
        description="Flatten a mixture of mixtures into leaf ingredients",
        input_descriptor=formulation,  # this is the formulation descriptor from the table
        output_descriptor=flat_formulation  # this is a new descriptor to represent the flattened formulation
    )

> Using the flattened formulation as an input, we create several featurizers to compute features that will be inputs to the machine learning model.

Reviewer: ```suggestion
Using the flattened formulation as an input, we create several "featurizers" to compute features; these will be the inputs to the machine learning model.
```

We create one predictor each for ingredient and label fractions, and two mean property predictors--
one that computes the mean price over all ingredients (this will be used to constraint the price of new margarita recipes) and one that computes the mean sucrose content of just the sweeteners.

.. code-block:: python

    from citrine.informatics.descriptors import RealDescriptor
    from citrine.informatics.predictors import LabelFractionsPredictor, IngredientFractionsPredictor, MeanPropertyDescriptor

    label_fractions_predictor = LabelFractionsPredictor(
        name="Label fractions",
        description="Total quantity that is from one of the component types",
        input_descriptor=flat_formulation,
        labels={"acid", "alcohol", "sweetener"},
    )

    ingredient_fractions_predictor = IngredientFractionsPredictor(
        name="Ingredient Fractions",
        description="Compute the fraction of each ingredient",
        input_descriptor=flat_formulation,
        ingredients=atomic_ingredients,

>     )

Reviewer 2: MAYBE - Should we say that this is just for demonstration purposes, and that in many cases using both these predictors can result in highly correlated inputs? I typically try to use either one or the other.

SHOULD - Even though this isn't full documentation, I **strongly** suggest we explain that the `atomic_ingredients` variable is a list of Material UIDs. This is often very confusing to users. They often try a list of Ingredient Names or a list of Material Names.

        predictor=ingredient_fractions_predictor, inputs=[flat_formulation],
    )
    price_descriptors = project.descriptors.from_predictor_responses(
        predictor=mean_price_predictor, inputs=[flat_formulation],
    )
    sweetness_descriptors = project.descriptors.from_predictor_responses(
        predictor=mean_sweetness_predictor, inputs=[flat_formulation]
    )

    blend_time = RealDescriptor("margarita~blend time", lower_bound=0, upper_bound=60, units="s")
    tastiness = RealDescriptor("margarita~tastiness", lower_bound=0, upper_bound=10, units="")

    ml_inputs = (
        ingredient_fractions_descriptors
        + label_fractions_descriptors
        + price_descriptors
        + sweetness_descriptors
        + [blend_time]
    )
    ml_model = AutoMLPredictor(
        name="ML model for tastiness",
        description="",
        output=tastiness,
        inputs=ml_inputs
    )

Where did the descriptor keys ``margarita~blend time`` and ``margarita~tastiness`` come from?
They came from concatenating the headers in the variables in the table, and the bounds and units came from the attribute templates.

> It's a lot to keep track of, we there is there :func:`~citrine.resources.descriptors.DescriptorMethods.descriptors_from_data_source` method.

Reviewer 2: MUST - Looks like a typo
> It's a lot to keep track of, which is why there is...

.. code-block:: python

    from citrine.informatics.predictors import GraphPredictor

    predictors_list = [
        simple_mixture_predictor,
        ingredient_fractions_predictor,
        label_fractions_predictor,
        mean_price_predictor,
        mean_sweetness_predictor,
        ml_model
    ]

    graph_predictor = GraphPredictor(
        name="Graphical model for tastiness of blended margaritas",
        description="",
        predictors=predictors_list,
        training_data=[data_source] # the data source we defined above
    )
    graph_predictor = project.predictors.register(graph_predictor)

One representation of this graphical model is shown below.

.. figure:: _static/formulations_graphical_model.png
    :align: center

    Graphical model to predict tastiness of a margarita

Defining a Design Space
-----------------------

Now that we have a trained predictor, the next step to finding the *world's best margarita* is to define the search space.
A :class:`~citrine.informatics.design_spaces.formulation_design_space.FormulationDesignSpace` defines the ingredients, the labels, and any constraints.

> In addition to the labels used in the predictor, we also create a label for "simple syrup A" and "simple syrup B."

Reviewer 2: MAYBE - Should we specify that this will only work if there are Ingredients with these labels already in the MH? I could see a lot of people getting to this stage and saying "Oh! I want to apply a new label. Why isn't it working?"

Author: This actually does not require that the labels exist in the material history. We are labeling the formulations emitted by the design space and constraining them directly, so there's no interaction with the predictor or the training data.

This allows us to specify that one and only one of the simple syrups should be used.

Notice that the design space's formulation descriptor corresponds to the original, unflattened formulation.
That's because we want to describe the margarita we are physically making in our kitchen.

> The predictor takes care of flattening it.

Reviewer 2: SHOULD - Again, add "it to its atomic ingredients." might be more clear for those not familiar with "flattening".

        labels={
            "acid": {"lime juice"},
            "alcohol": {"tequila", "triple sec"},
            "sweetener": {"triple sec"},
            "simple syrup": {"simple syrup A", "simple syrup B"}
        },
        constraints={
            IngredientCountConstraint(formulation_descriptor=formulation, min=3, max=5),
            IngredientCountConstraint(formulation_descriptor=formulation, min=1, max=1, label="simple syrup"),
            IngredientFractionConstraint(formulation_descriptor=formulation, ingredient="lime juice", min=0.15, max=0.30)
        }
    )

But that's not all; we also have to define the ``blend time`` input.
We do this by wrapping a :class:`~citrine.informatics.design_spaces.product_design_space.ProductDesignSpace` around the formulation design space and a ``blend time`` dimension.

.. code-block:: python

    from citrine.informatics.design_spaces import ProductDesignSpace

    design_space = ProductDesignSpace(
        name="margaritas design space",
        description="",
        subspaces=[fds],
        dimensions=[ContinuousDimension(descriptor=blend_time, lower_bound=5.0, upper_bound=20.0)]
    )

>     design_space = project.design_spaces.register(design_space)

Reviewer 2: MAYBE - do we want to mention that we can also use enumerated dimensions here, even if we don't demonstrate it? A lot of experimental folks are interested in those as well.

Author: Using an enumerated design space would require an `IngredientsToSimpleMixturePredictor`, and that's not included in this write-up since it's not being brought to beta yet.

Proposing New Formulation Candidates
------------------------------------

With the pieces assembled, we define a :class:`~citrine.informatics.workflows.design_workflow.DesignWorkflow`.
Our goal is to find the margarita that is most likely to have a tastiness score above 7.5, while keeping the cost per kg below $12.50.
We define an :class:`~citrine.informatics.scores.LIScore` with this objective and constraint, define a design workflow with the predictor and design space, and trigger the design workflow on the score.

.. code-block:: python

    from citrine.informatics.scores import LIScore
    from citrine.informatics.objectives import ScalarMaxObjective
    from citrine.informatics.constraints import ScalarRangeConstraint
    from citrine.informatics.workflows.design_workflow import DesignWorkflow

    score = LIScore(
        objectives=[ScalarMaxObjective(descriptor_key=blend_time.key)],
        baselines=[7.5],
        constraints=[ScalarRangeConstraint(
            descriptor_key=price_descriptors[0].key, upper_bound=12.50, lower_bound=0.00
        )]
    )

    design_workflow = DesignWorkflow(

>         name="best margarita below 12.50",

Reviewer 2: SHOULD - we should probably not include targets in the name of the design workflow, since they are independent and may confuse people into making a new design workflow every time they want to change targets.

A formulation is, generally speaking, a mixture of mixtures.
Individual ingredients are mixed together to form more complex ingredients, which may themselves be mixed with other ingredients and/or mixtures, and so on.
Generally, the top-level formulation has some properties that we wish to optimize.
The ingredients themselves may also have known properties that are useful for understanding the properties of the top-level formulation.
In addition, each mixing process may be characterized by attributes such as the temperature of a reaction vessel or the speed of a mixing element.

The GEMD data model provides a way to record all of the information relevant to a formulations problem,
and Citrine's AI Engine turns this information into trained models that can predict the properties of novel formulations.
This section demonstrates how to express a complete formulations example in Citrine-python, from data ingestion to new candidate generation.

Example Raw Data
----------------

To demonstrate the formulations machinery, we consider a toy problem -- margaritas!
We scour our pantry and gather some materials, laid out in the table below.

> These are known as "atomic" materials because for our purposes they are not broken into more fundamental materials.

Author: I don't think we use any term for this concept anywhere else. The docstring for `SimpleMixturePredictor` just says "flattens to a simple mixture" without explaining what makes it simple.

I like "atomic" over "raw", since the primary quality is that we are uninterested in any smaller components.

Make sure to set the bounds wide enough to encompass all anticipated use cases of the templates.

.. Attention:: At this time, currency units are not supported by default because of issues around fluctuating conversion rates.
    For the "price" template, we therefore use units of 1/kg instead of $/kg.

.. note:: In the code examples that follow, imports are assumed to be cumulative.
    Once an import has been demonstrated, we do not include it in subsequent code blocks.

.. code-block:: python

    from citrine.resources.parameter_template import ParameterTemplate
    from citrine.resources.property_template import PropertyTemplate
    from gemd.entity.bounds import RealBounds

    # Assume we have already created a project and dataset, named "project" and "dataset."
    # We give the templates ids with a custom scope so that they can be easily retrieved.
    blend_time_template = dataset.parameter_templates.register(
        ParameterTemplate(
            "blend time",
            bounds=RealBounds(0, 60.0, "s"),
            uids={"margaritas-id": "blend time template"}
        )

>     )

Author: I'd like this code to be as brief and efficient as possible, since learning about GEMD isn't the goal here.

This assumes that the material specs for the atomic materials and the simple syrups have already been uploaded.

.. Attention:: It is important to assign human-readable uids to the Material Runs, as these will be the unique identifiers in the resulting table.

.. code-block:: python

    from citrine.resources.process_spec import ProcessSpec
    from citrine.resources.ingredient_spec import IngredientSpec
    from citrine.resources.material_spec import MaterialSpec
    from gemd.entity.value import NominalReal
    from gemd.entity.link_by_uid import LinkByUID

    mix_margarita_spec = dataset.process_specs.register(
        ProcessSpec("mix margarita B", template=LinkByUID("margaritas-id", "mix template"))
    )
    dataset.ingredient_specs.register(
        IngredientSpec(
            "simple syrup",
            material=simple_syrup_B_spec,  # assume that this and the other relevant specs exist in memory
            process=mix_margarita_spec,
            labels=["simple syrup"]
            mass_fraction=NominalReal(nominal=0.15, units="")
        )
    )
    dataset.ingredient_specs.register(
        IngredientSpec(

>             "triple sec",

Author: We already use "lime juice" to draw a distinction between an ingredient name ("lime juice") and a material identifier ("bottled lime juice"). Also drawing the distinction here doesn't add anything

Finally, run objects are created corresponding to each spec, to represent what actually happened.
Whew!
That's a lot, which is why this is best done with additional tooling, whether in code or in the GUI.

A rendering of this example material history is shown below.

.. figure:: _static/GEMD_history_example.png
    :align: center

    Material History for Blended Margarita B

Repeating this process once for each margarita sample, we can build up a rich dataset for machine learning.

Building a Table
----------------

We now build a GEM Table to represent the margaritas' material histories in tabular format.
This table will be used as training data when building a machine learning graphical model.
For more detailed information on GEM Tables, see the section on :doc:`data extraction <data_extraction>`.

In order to make a GEM Table, we start with a Table Configuration object.
In this example we will build up the configuration in small steps.
As we will see, templates are crucial to configuring the table.

The code below defines the rows and defines one column that contains the identifier of each row.

.. code-block:: python

>     from citrine.gemtables.rows import MaterialRunByTemplate

Author: why remove MaterialRunByTemplate?

------------------------------------

With the pieces assembled, we define a :class:`~citrine.informatics.workflows.design_workflow.DesignWorkflow`.
Our goal is to find the margarita that is most likely to have a tastiness score above 7.5, while keeping the cost per kg below $12.50.
We define an :class:`~citrine.informatics.scores.LIScore` with this objective and constraint, define a design workflow with the predictor and design space, and trigger the design workflow on the score.

.. code-block:: python

    from citrine.informatics.scores import LIScore
    from citrine.informatics.objectives import ScalarMaxObjective
    from citrine.informatics.constraints import ScalarRangeConstraint
    from citrine.informatics.workflows.design_workflow import DesignWorkflow

    score = LIScore(
        objectives=[ScalarMaxObjective(descriptor_key=blend_time.key)],
        baselines=[7.5],
        constraints=[ScalarRangeConstraint(
            descriptor_key=price_descriptors[0].key, upper_bound=12.50, lower_bound=0.00
        )]
    )

    design_workflow = DesignWorkflow(
        name="best margarita",
        design_space_id=design_space.uid,
        predictor_id=graph_predictor.uid,

>         processor=None  # we use the default continuous search processor

Reviewer: ```suggestion
        processor_id=None  # we use the default continuous search processor
```

            allowed_labels=["sweetener", "acid", "alcohol", "simple syrup"]
        )
    )

.. Hint:: Here we must distinguish between "name" and "identifier".
    "Name" in this case is an abstract way of describing the types of things that go into a process.
    "Identifier" is the unique way of referring to one particular material.
    In this example the template allows the name "lime juice," but does not specify the specific material.
    For simplicity, this example only includes one type of lime juice, "bottled lime juice," but we might also have "fresh-squeezed lime juice" or several batches of bottled lime juice, each with their own unique identifier.
    In some cases, such as "ice," for which we know that only one material will be used, we have conflated the name and identifier.

    Specifying ``allowed_names`` is not required for GEMD, but will make it easier to build the eventual table.
    In this case, the table will have a super-column describing the use of lime juice in the margarita, with sub-columns for the identifier of the specific lime juice used, its quantity, and any labels to apply.

>     If a suitable set of names are not known, it is possible to be generic: "mixing input 1," "mixing input 2," etc.

Reviewer 2: This is so much more clear!

    formulation = FormulationDescriptor("mixed and blended margarita")
    data_source = GemTableDataSource(table_id=table.uid, table_version=table.version, formulation_descriptor=formulation)

The first component of the graphical model is a :class:`~citrine.informatics.predictors.simple_mixture_predictor.SimpleMixturePredictor`, which flattens the input formulation--it repeatedly replaces components with their ingredients until only the atomic ingredients remain.
This flattening efficiently teaches the predictor about the relationship between materials.
In this case, it learns exactly how "simple syrup A" and "simple syrup B" are similar to each other because they both contain atomic materials sugar and water but in slightly different amounts.
Although the homogeneous representation is not entirely appropriate for all formulations problems, it is usually an excellent approximation,

> especially when coupled with flexible machine learning models that can emulate more complexsubtle relationships within the data.

Reviewer 2: ```suggestion
especially when coupled with flexible machine learning models that can emulate more complex relationships within the data.
```
